### DISCUSSION TITLE
Requests distribution

### DISCUSSION THREAD

**Discussion by SlavaSkvortsov** [Kludex/uvicorn#2467] [Q&A] [answered]
Hey 👋 

Recently we noticed that request distribution between different processes on a single machine looks like this:
<img width="943" alt="image" src="https://github.com/user-attachments/assets/3a29fb4a-73d1-4647-b9f5-32572081d794">
Some of the processes get most of the requests, while others barely pick up any. 

We use `uvicorn` together with `supervisor` ([guide](https://www.uvicorn.org/deployment/#supervisor)). 

<details>
  <summary>Here is an example of the `supervisor` config:</summary>
  
```
[fcgi-program:uvicorn]
numprocs=4
command=uvicorn app:app --fd 0
socket=tcp://127.0.0.1:8000
process_name=test_%(process_num)02d
```
  
</details>

<details>
  <summary>and here is a simple app for testing</summary>
  
```python3
import asyncio
import os

from fastapi import FastAPI

app = FastAPI()

# Global counter and lock
request_count = 0
request_count_lock = asyncio.Lock()


# Function to safely increment and print request count
async def increment_request_count():
    global request_count
    async with request_count_lock:
        request_count += 1


@app.get("/")
async def read_root():
    await increment_request_count()
    return {"pid": os.getpid(), 'count': request_count}

```
  
</details>

<details>
  <summary>and a simple script to test it</summary>
  
```python3
import asyncio

import httpx
from httpx import Limits


async def call(client):
    response = await client.get('http://127.0.0.1:8000')
    print(response.text)
    return response.json()


async def main():
    results = {}

    async with httpx.AsyncClient(limits=Limits(max_connections=10, max_keepalive_connections=10)) as client:
        for result in (await asyncio.gather(*[call(client) for _ in range(1000)])):
            results[result['pid']] = result['count']

    print(results)


if __name__ == '__main__':
    asyncio.run(main())

```
  
</details>

This setup provides poor requests distribution. I tried using [uvicorn as a process manager](https://www.uvicorn.org/deployment/#built-in), but the result is the same.

Previously, there was a [similar discussion](https://github.com/encode/uvicorn/discussions/1138), but about `gunicorn`. Right now, `gunicorn` [struggles](https://github.com/benoitc/gunicorn/pull/2938) with the same problem, but there is a PR that fixes it by utilizing [SO_REUSEPORT](https://man7.org/linux/man-pages/man7/socket.7.html) socket option that 

> For TCP sockets, this option allows accept(2) load distribution in a multi-threaded server to be improved by using a distinct listener socket for each thread.  This provides improved load distribution as compared to traditional techniques such using a single accept(2)ing thread that distributes connections, or having multiple threads that compete to accept(2) from the same socket

Sorry for the long intro 😅 Here is my question:
Is there a way to achieve more or less even request distribution across the workers on the same machine using `uvicorn`?

If it's not possible to do that using just `uvicorn`, which approach could you recommend to improve the balancing? 

Thank you!!

**Comment by Kludex** [comment]
Hi @SlavaSkvortsov , right... I don't know where the issue relies on. We do set the `SO_REUSEPORT`, but it doesn't seem to be enough:

https://github.com/encode/uvicorn/blob/a50753268a9033dcdc57d6dee45f3829cdedf8f0/uvicorn/config.py#L512

Any ideas @graingert @abersheeran ?

My recommendation would be to rely on an external load balancer e.g. nginx.

**Reply by graingert** [reply]
It looks like you're not using SO_REUSEPORT and from my reading of the docs and the way to use it is to call socket.bind() on each subprocess rather than sharing a socket and competing for accept()

WinSock seems to behave differently though and I don't understand what the difference is will need an experiment

Notice how in the gunicorn PR the sockets are created for each worker https://github.com/benoitc/gunicorn/pull/2938/files#diff-2cb7ac1e139f4df283747ee8c3cd0f26bfaa8aaf4c5891097f8c0e382d1a618aR585

**Reply by graingert** [reply]
Here's a clearer explanation https://lwn.net/Articles/542629/

**Reply by Kludex** [reply]
Thanks Thomas.

**Reply by graingert** [reply]
Also I think there's a tradeoff involved, where requests will be handled more slowly at the expense of more even load. Eg when one server process is ready to handle a connection but the connection is earmarked for a busy server then the request will be delayed

**Reply by graingert** [reply]
https://man.freebsd.org/cgi/man.cgi?setsockopt(2) bsd is different and  I think needs SO_REUSEPORT_LB which isn't supported in python

**Reply by graingert** [reply]
Also there's an additional footgun here, I think you should not set REUSEPORT when running the dev server (reload mode) otherwise if a user runs the dev server on two different terminals Linux will load balance between them, and in reality the user will expect the second server to fail to bind to an already in use port

**Reply by Kludex** [reply]
> [man.freebsd.org/cgi/man.cgi?setsockopt(2)](https://man.freebsd.org/cgi/man.cgi?setsockopt(2)) bsd is different and I think needs SO_REUSEPORT_LB which isn't supported in python

I've found this implementation around:

https://github.com/cherrypy/cheroot/blob/1ff20b18e98e424e0616654c28c92524deef348a/cheroot/server.py#L1983-L2007

**Reply by graingert** [reply]
But I don't think python socket defines that flag

**Reply by graingert** [reply]
see how SO_REUSEPORT behaves differently on macos and linux: 

```python
import threading
import sys
import socket
import concurrent.futures

def main():
    serving = threading.Event()
    accepted = set()

    def serve(name):
        with socket.create_server(("localhost", 8080), reuse_port=True) as sock:
            serving.set()
            while True:
                client_sock, address = sock.accept()
                accepted.add(name)
                if len(accepted) > 1:
                    print("done!")
                    return
                print(f"sock accepted on server {name}")
                with client_sock:
                    pass

    def client():
        serving.wait()
        while True:
            with socket.create_connection(("localhost", 8080)):
                pass

    with concurrent.futures.ThreadPoolExecutor(3) as tpe:
        f1 = tpe.submit(serve, 1)
        f2 = tpe.submit(serve, 2)
        f3 = tpe.submit(client)
        for v in concurrent.futures.as_completed((f1, f2, f3)):
            v.result()



if __name__ == "__main__":
    sys.exit(main())
```

on macos this runs forever printing sock accepted on server 1 forever, on linux it terminates after both servers accept a connection

**Reply by graingert** [reply]
one neat thing you can do is for the first socket bind you can set SO_REUSEPORT 0, then after it's bound set SO_REUSEPORT 1, this way if another server is bound to the port you get an OSError 98

**Comment by abersheeran** [comment]
Please make sure you use the new multi-process manager and increase the concurrency (the test script you gave establishes at most 10 connections, so the test is too accidental).

**Comment by minmax** [comment]
We have the same problem with gunicorn, our application structure is an asynchronous gunicorn-worker with processing of most requests in a synchronous thread (there is only one thread). `gunicorn -> uvicorn-worker -> sync thread`

Perhaps if you repeat this load profile, the balancing problem will be visible.

**Comment by graingert** [comment, marked as answer]
@SlavaSkvortsov does https://github.com/encode/uvicorn/pull/2472/files fix it for you?

**Reply by SlavaSkvortsov** [reply]
First of all, wanna say huge thanks for working on it 🙏 Very appreciate it! 

I was surprised to learn that it works differently on MacOS and Linux, so when I was testing locally (on MacOS), I was able to easily reproduce the issue with the script I shared in the first message.

<details>
  <summary>My OS version</summary>
  
```
➜  ~ sw_vers
ProductName:		macOS
ProductVersion:		14.3
BuildVersion:		23D56  
```

</details>

 However, when I read the thread and understood that there was a difference, I went to our server to check it there.

<details>
  <summary>Server configuration is</summary>
  
```
~ $ cat /etc/os-release
NAME=Gentoo
ID=gentoo
PRETTY_NAME="Gentoo Linux"
ANSI_COLOR="1;32"
HOME_URL="https://www.gentoo.org/"
SUPPORT_URL="https://www.gentoo.org/support/"
BUG_REPORT_URL="https://bugs.gentoo.org/"
VERSION_ID="2.14"
```
  
</details>

With the old script all was good and I could not spot any issues, but I saw them in our metrics. We changed the script a bit to reproduce the issue

<details>
  <summary>New test script</summary>
  
```python
import asyncio
from statistics import mean, stdev

import httpx
from httpx import Limits


async def call():
    async with httpx.AsyncClient(timeout=60, limits=Limits(max_connections=100, max_keepalive_connections=10)) as client:
        response = await client.get('http://127.0.0.1:8000/')
        return response.json()


async def main():
    results = {}

    for result in (await asyncio.gather(*[call() for _ in range(1000)])):
        pid, count = result['pid'], result['count']
        results[pid] = max(results.get(pid, 0), count)

    print(results)
    print(stdev(results.values()) / mean(results.values()) * 100)


if __name__ == '__main__':
    asyncio.run(main())
```
  
</details>


With this script I can still see the issue on both Gentoo and MacOS even with the fix you shared :( 

Want to mention, that for MacOS the issue is more severe. For example, a typical run for MacOS looks like 

```
{9236: 491, 9233: 337, 9235: 81, 9234: 1}
```

And for Gentoo it's 

```
{133531: 413, 133533: 287, 133532: 129, 133530: 171}
```

**Reply by graingert** [reply]
Are you still running with supervisor? You need to run with uvicorn's multiprocessing manager and --socket-load-balance=True

Bear in mind macos didn't support socket load balancing so it won't work there

**Reply by graingert** [reply]
I also just noticed I got the logic inverted of the supported modes and have pushed a fix and test

**Reply by SlavaSkvortsov** [reply]
Oh wow, it's working! 

On Gentoo:
Patched, with supervisor:
```bash
(venv-patched) ~/slava.skvortsov $ python test-copy.py 
{80879: 324, 80881: 341, 80880: 75, 80882: 260}
48.70701523736938
```

Patched, with uvicorm's process manager
```bash
(venv-patched) ~/slava.skvortsov $ python test-copy.py 
{87512: 279, 87511: 251, 87510: 226, 87509: 244}
8.80605852051113
```

Amazing! 

I don't understand, though, why it's not working with supervisor? Is there a way to make them compatible?

**Reply by graingert** [reply]
supervisor binds the socket and passes it to the child process. The fix should be in supervisor to bind a new socket for each process with SO_REUSEPORT

**Reply by SlavaSkvortsov** [reply]
I see, thanks! 

Do you plan to release the PR? Would like to avoid using a fork in production 😅

**Reply by graingert** [reply]
currently I'm fighting coverage issues, and don't have much time to work on it - I'll need someone to finish it off for me.

I raised an issue with supervisor to add the feature there: https://github.com/Supervisor/supervisor/issues/1661

**Comment by mdpdesign** [comment]
Hi, I stumbled upon this discussion as I'm also experiencing such problem when having multiple workers in `uvicorn`, after some initial load only 1 worker is serving responses to the client and other workers are idling. I'm struggling to understand what exactly is happening and how to solve the bottleneck. 

The FastAPI app is rather simple, I have few endpoints that are talking to DB querying some products - I'm using MSSQL Express 2019 (as this is a requirement..). Currently I'm running the app inside WSL2. 

I'm trying to do my best to have everything `async`, so I use `SQLAlchemy` with `aiodbc` + async session and engine + fastapi-pagination for cursor based pagination for the results from DB, to limit the number of results returned, the endpoint function looks like this:

```python
@products_router.get(
    "/",
    summary="Get list of all products, paginated",
)
async def get_all_products(
    session: DBEngineDep,
    params: TotalCursorParams = Depends(),
) -> TotalCursorPage[Product]:
    """
    Get list of all products, paginated
    """

    all_products: TotalCursorPage[Product] = await insertgt.get_all_products(
        session=session,
        params=params,
    )

    if all_products.total is not None and all_products.total > 0:
        return all_products
    else:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="No products found",
        )
```

And a single call to it takes ~400ms - 500ms:

```
curl -s -o /dev/null -w "%{http_code} %{time_total} %{http_version} %{method} %{url}\n" -X 'GET' 'http://localhost:8000/api/v1/products/?size=500' -H 'accept: application/json'

200 0.432331 1.1 GET http://localhost:8000/api/v1/products/?size=500
200 0.421117 1.1 GET http://localhost:8000/api/v1/products/?size=500
200 0.403372 1.1 GET http://localhost:8000/api/v1/products/?size=500
```

I start the application from `main.py` programmatically:

```python
if __name__ == "__main__":
    workers: int = TypeAdapter(int).validate_python(os.getenv("WORKERS", 6))
    reload: bool = TypeAdapter(bool).validate_python(os.getenv("RELOAD", False))

    uvicorn.run(
        app="main:app",
        host="0.0.0.0",
        port=8000,
        reload=reload,
        workers=workers,
        # uvloop works only on Linux!
        loop="uvloop",
        log_config="log-config.yaml",
        log_level="debug",
    )
```

I'd expect that if I start the app with 6 workers, all of them would pick up the requests and would load balance equally. But unfortunately, this isn't the case. I test the app with `Locust` and 10 users in parallel to query all the DB items - basically to know how fast a user can query all items in the DB. 

After initial load I see usually ~3-4 out of 6 workers being active and serving responses, then after some of the users finish querying the DB, all other users are served just by 1 `uvicorn` worker which greatly impacts the app performance.

I must add that when I issue another request "on the side" (outside Locust tests) app is still responsive and I get the response back - so I think `async` works.

Here is very trivial example of this behavior - 2 first users queried DB in ~1 min., then we observe huge bottleneck (1 worker serving requests, others doing nothing):

```
[2025-02-14 15:33:21,856] DESKTOP-GMSMBCO/INFO/locustfile: User 'Ethan   ', elapsed time: 58.048524 (0:00:58.048524) seconds
[2025-02-14 15:33:22,053] DESKTOP-GMSMBCO/INFO/locustfile: User 'Juan    ', elapsed time: 58.245155 (0:00:58.245155) seconds
[2025-02-14 15:35:28,922] DESKTOP-GMSMBCO/INFO/locustfile: User 'Christian', elapsed time: 185.110893 (0:03:05.110893) seconds
[2025-02-14 15:35:30,426] DESKTOP-GMSMBCO/INFO/locustfile: User 'Daniel  ', elapsed time: 186.615245 (0:03:06.615245) seconds
[2025-02-14 15:36:46,746] DESKTOP-GMSMBCO/INFO/locustfile: User 'Benjamin', elapsed time: 262.935235 (0:04:22.935235) seconds
[2025-02-14 15:36:47,084] DESKTOP-GMSMBCO/INFO/locustfile: User 'Andrew  ', elapsed time: 263.275920 (0:04:23.275920) seconds
[2025-02-14 15:37:43,826] DESKTOP-GMSMBCO/INFO/locustfile: User 'Jacob   ', elapsed time: 320.013285 (0:05:20.013285) seconds
[2025-02-14 15:37:45,825] DESKTOP-GMSMBCO/INFO/locustfile: User 'Zachary ', elapsed time: 322.013534 (0:05:22.013534) seconds
[2025-02-14 15:37:53,055] DESKTOP-GMSMBCO/INFO/locustfile: User 'Dennis  ', elapsed time: 329.245970 (0:05:29.245970) seconds
[2025-02-14 15:37:53,271] DESKTOP-GMSMBCO/INFO/locustfile: User 'John    ', elapsed time: 329.460643 (0:05:29.460643) seconds
```

I observed that when I kill the currently struggling serving worker process, other workers are picking up the load and are starting to serve responses - so RPS increases for some time, then the situation repeats.

Most probably I will have to deploy this app (stack) in a container running on a PC not as a web app (k8s etc.), so running it with multiple healthy workers makes sense to me

I prepared a video which demonstrates this issue, I will try to attach it

Test & app startup:

https://github.com/user-attachments/assets/c7bf75ed-c149-4859-83ab-e141bbb5e20d

App initially serving requests with few workers, then bottleneck starts:

https://github.com/user-attachments/assets/c86a2506-04f8-42ab-9404-f61afe01f778

I had to kill "the one" worker process and other processes started serving again, but not all of them:

https://github.com/user-attachments/assets/4f00f809-338a-4bb1-b96c-bf43d03165cb

Issue repeats:

https://github.com/user-attachments/assets/4556d4ee-175a-43db-8d4d-129e9af6c855

**Reply by graingert** [reply]
This should be fixed by my PR currently need a decision on using pytest-cov for coverage of subprocess

**Reply by mdpdesign** [reply]
@graingert - thanks for the response! 

I tried installing `uvicorn` with `uv` (`uv pip install "git+https://github.com/encode/uvicorn@socket-load-balance"`) from your branch, but doing almost the same test didn't help (maybe I'm doing something wrong 😅) - I still observed only 1 worker serving responses after a while

The difference was that I started the app with `uvicorn --socket-load-balance --workers 6 --log-config ../../log-config.yaml --log-level debug "main:app"` instead calling `main.py`

I must add that I also tried serving the app with `gunicorn` & uvicorn workers - `uv run gunicorn main:app -w 6 --reuse-port -k uvicorn.workers.UvicornWorker` - but that didn't help as well 🙁

**Reply by graingert** [reply]
Can you make a runnable example, I can download and run? Can you try periodically dumping the tracebacks from `sys._current_frames()` and inspecting the tracebacks when the worker process is stalled.

Does this problem happen with gunicorn?

Do all the workers stall at the same time (except one)?

**Comment by mdpdesign** [comment]
@graingert answering your questions:

> Does this problem happen with gunicorn?

Yes, I tested the real app with `gunicorn` with `uvicorn` worker class and it behaved the same way

> Do all the workers stall at the same time (except one)?

I observed that when I have 10 users trying to get all DB items, it feels like some users becomes tied to a particular worker, when user finishes all requests, worker stops doing the work. Users that left, queue on a single worker, until I kill that worker process. It's more visible in the videos I attached

I think I've been able to create a mock app that behaves similarly to the real app. The key is to have some CPU load generated, initially I tried just simple `await asyncio.sleep(0.5)` - to simulate DB query as per my previous answer. But that didn't show the issue

Here's the mock app:

[uvicorn-workers-troubleshooting-app.tar.gz](https://github.com/user-attachments/files/18809365/uvicorn-workers-troubleshooting-app.tar.gz)

I used `uv` to manage the project, to run it:

```
# Create .venv and install packages
uv sync --python 3.13.2

# Run application
uv run src/app/main.py

# Run locust for load testing
uv run locust -f locustfile.py --web-host 0.0.0.0 --web-port 8090 -u 10 -r 10 --tags all-products |& tee locust-log.log

# Logs included from my own testing (correlates to videos)
original-app-logs.log
original-locust-log.log
```

And the recording that I did, (I had to split it into ~10MB files)


https://github.com/user-attachments/assets/6d26e82a-938e-4e6b-a04f-31ce6f8da75e

In the below video around 0:05 (5 sec.), we can see that as soon as 3 users finish their requests - workers become idle, not picking up any requests even though there are still 7 users left to be served, we see that only 2 workers are serving the responses for a while. Then around 1:21 another user finished, and only 1 worker is still serving responses

https://github.com/user-attachments/assets/e24b803c-d38d-4588-89e9-a13b368fd68f

In the below video, around 0:06 I killed that one busy worker, and 4 workers picked up the work - still not all of them (app starts with 6 workers). Then around 0:50 problem repeats

https://github.com/user-attachments/assets/d9f9482f-d661-4b57-8f8f-cd4c8958cb03

This issue also happens on a MacOS - I used the same example app that is attached:

![001-initial-load](https://github.com/user-attachments/assets/3d9ccada-69d0-4b2f-bdae-c3a4bf641177)
![002-3users-done](https://github.com/user-attachments/assets/0b79f8bd-a176-4ea3-beb5-03a01482707f)
![003-degraded-perf](https://github.com/user-attachments/assets/82c10d54-d8e8-42ff-981f-9f23d91e5456)
![004-degraded-perf-locust](https://github.com/user-attachments/assets/cba560d4-cb3d-4006-bb14-c55c747aabb0)

I hope this helps

**Reply by mdpdesign** [reply]
Hi @Kludex & @graingert any thoughts from your side on the provided information?

**Reply by graingert** [reply]
Can you link to the mock app on GitHub, and produce a tool that makes requests to the mock app and demonstrates the requests blocking completely

**Reply by mdpdesign** [reply]
Sure, here's the repo with that example app + there is a `locustfile.py` file to do load testing with [Locust load testing tool](https://locust.io/) - https://github.com/mdpdesign/uvicorn-workers-troubleshooting-app

**Comment by methane** [comment]
If 16 requests are randomly distributed among 8 workers, there is approximately an 80% chance that at least one worker will receive 4 or more requests.

Therefore, to improve the 99th percentile latency and use CPU efficiently, requests should be distributed using better algorithms such as least connections rather than randomly.

Load balancers such as nginx, HAProxy, and Envoy can distribute requests using least connections or round-robin methods. For this purpose, WSGI/ASGI servers need to have each process listen on independent ports.

I had Codex implement it to see how complex supporting multiple ports would be.
To avoid creating noise for developers, I created the PR within a fork.

https://github.com/methane/uvicorn/pull/1
